File: cpp/141. Linked List Cycle.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
// Floyd’s algorithm, Time: O(n), Space: O(1)
public:
    bool hasCycle(ListNode *head) {
        ListNode *slow = head, *fast = head;       
        while(fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            if(fast==slow) return true;
        }
        return false;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}

File: python/141-Linked List Cycle.py
from linklist_helper import *

class Solution:
    # def hasCycle(self, head: ListNode) -> bool:
    #     "set, Time: O(n), Space: O(n)"
    #     if head is None or head.next is None: return False
    #     check_dup = set()
    #     while head:
    #         if head in check_dup:
    #             return True
    #         else:
    #             check_dup.add(head)
    #             head = head.next
    #     return False
    def hasCycle(self, head: ListNode) -> bool:
        "Slow Fast Pointer, Time: O(n), Space: O(1)"
        if head is None or head.next is None: return False
        slow, fast = head, head
        while True:
            if fast is None or fast.next is None:
                return False
            slow = slow.next
            fast = fast.next.next
            if fast == slow:
                return True
